### PR TITLE
Update matches.js

### PR DESCRIPTION
**Pull request by tech62**
Modification to use team.name in replacement of match.team1_string
Adding server name in matches/mymatches views
Lot of SQL request updated, please check before merge !

### CODE REVIEW

File: routes/matches/matches.js [hunk 1]
```diff
@@ -329,14 +329,19 @@ router.get("/", async (req, res, next) => {
     let isAscending = req.query?.asc == null ? false : req.query.asc;
     let sql = 
       "SELECT mtch.id, mtch.user_id, mtch.server_id, mtch.team1_id, mtch.team2_id, mtch.winner, mtch.team1_score, " +
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:331]: Will these joins not fail if a team gets deleted? I would much rather have a check to see if `is_pug` (as teams are removed after a pug has finished). is enabled in the first query and just use the default statement, otherwise maybe use a join.

Also, try not to bring in everything within a match either. If you need more information, you can just make an API call specific to that route, such as map stats. The whole point of the routes is trying to make the calls as independent as possible.

File: routes/matches/matches.js [hunk 2]
```diff
@@ -392,10 +397,13 @@ router.get("/mymatches", Utils.ensureAuthenticated, async (req, res, next) => {
               "JOIN user usr ON mtch.user_id = usr.id " +
               "WHERE user_id = ? ORDER BY id DESC";*/
       let sql = "SELECT mtch.id, mtch.user_id, mtch.server_id, mtch.team1_id, mtch.team2_id, mtch.winner, mtch.team1_score, " +
-      "mtch.team2_score, mtch.team1_series_score, mtch.team2_series_score, mtch.team1_string, mtch.team2_string, " +
+      "mtch.team2_score, mtch.team1_series_score, mtch.team2_series_score, team1.name as team1_string, team2.name as team2_string, " +
       "mtch.cancelled, mtch.forfeit, mtch.start_time, mtch.end_time, mtch.max_maps, mtch.title, mtch.skip_veto, mtch.private_match, " +
-      "mtch.enforce_teams, mtch.min_player_ready, mtch.season_id, mtch.is_pug, usr.name as owner, mp.team1_score as team1_mapscore, mp.team2_score as team2_mapscore " +
+      "mtch.enforce_teams, mtch.min_player_ready, mtch.season_id, mtch.is_pug, usr.name as owner, mp.team1_score as team1_mapscore, mp.team2_score as team2_mapscore, gs.display_name " +
       "FROM `match` mtch JOIN user usr ON mtch.user_id = usr.id LEFT JOIN map_stats mp ON mp.match_id = mtch.id " +
+      "LEFT JOIN team team1 ON team1.id = mtch.team1_id " +
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:404]: Same as above.

File: routes/matches/matches.js [hunk 3]
```diff
@@ -460,14 +468,28 @@ router.get("/:match_id", async (req, res, next) => {
       req.user !== undefined &&
       (matchRow[0].user_id == req.user.id || Utils.superAdminCheck(req.user))
     ) {
-      sql = "SELECT * FROM `match` where id=?";
+//      sql = "SELECT * FROM `match` where id=?";
+	sql =  "SELECT `match`.id, `match`.user_id, server_id, team1_id, team2_id, winner, " +
+        "team1_score, team2_score, team1_series_score, team2_series_score, " +
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:473]: Same as above. Calls should be separated.

File: routes/matches/matches.js [hunk 4]
```diff
@@ -460,14 +468,28 @@ router.get("/:match_id", async (req, res, next) => {
...
+        "season_id, is_pug, map_sides, gs.display_name FROM `match`" +
+        "LEFT JOIN team team1 ON team1.id = `match`.team1_id " +
+        "LEFT JOIN team team2 ON team2.id = `match`.team2_id " +
+        "LEFT JOIN game_server gs on gs.id = match.server_id " +
+        "where `match`.id = ?";
+
     } else {
       sql =
-        "SELECT id, user_id, server_id, team1_id, team2_id, winner, " +
+        "SELECT `match`.id, `match`.user_id, server_id, team1_id, team2_id, winner, " +
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:484]: If using joins, please try to follow convention as well. `SELECT ... FROM `match` as mtch`.

File: routes/matches/matches.js [hunk 5]
```diff
@@ -525,24 +547,41 @@ router.get("/:match_id", async (req, res, next) => {
       req.user !== undefined &&
       (matchRow[0].user_id == req.user.id || Utils.superAdminCheck(req.user))
     ) {
-      sql = "SELECT * FROM `match` where id=?";
+//      sql = "SELECT * FROM `match` where id=?";
+	sql =
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:551]: Same as above, what will happen if all values in the team is deleted? Will this just return null? What if it's in the match table (per how pugs operate) compared to null data?

File: routes/matches/matches.js [hunk 6]
```diff
@@ -525,24 +547,41 @@ router.get("/:match_id", async (req, res, next) => {
...
+        "max_maps, title, skip_veto, private_match, enforce_teams, min_player_ready, " +
+        "season_id, is_pug, map_sides " +
+        "FROM `match`" +
+	"LEFT JOIN team team1 ON team1.id = `match`.team1_id " +
+        "LEFT JOIN team team2 ON team2.id = `match`.team2_id " +
+	"where `match`.id = ?";
     } else {
       sql =
-        "SELECT id, user_id, server_id, team1_id, team2_id, winner, " +
+	"SELECT `match`.id, `match`.user_id, server_id, team1_id, team2_id, winner, " +
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:563]: Same as above.

File: routes/matches/matches.js [hunk 7]
```diff
@@ -525,24 +547,41 @@ router.get("/:match_id", async (req, res, next) => {
...
+    if (!matches.length) {
+      res.status(404).json({ message: "No match found." });
+      return;
+    }
     res.set({
       "Cache-Control": "no-cache",
       "Connection": "keep-alive",
-      "Content-Type": "text/event-stream",
-      "X-Accel-Buffering": "no"
+      "Content-Type": "text/event-stream"
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:584]: Removal of `"X-Accel-Buffering": "no"` will cause event streams to not work with NGINX. Please put this back.

File: routes/matches/matches.js [hunk 8]
```diff
@@ -606,12 +645,14 @@ router.get("/:match_id", async (req, res, next) => {
       "FROM `match_pause` " +
       "WHERE match_id = ? ";
     let matchPauseInfo = await db.query(sql, matchId);
-
+    if (!matchPauseInfo.length) {
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:648]: Don't do this as connections will drop immediately instead of waiting for event changes.

File: routes/matches/matches.js [hunk 9]
```diff
@@ -525,24 +547,41 @@ router.get("/:match_id", async (req, res, next) => {
...
+        "LEFT JOIN team team1 ON team1.id = `match`.team1_id " +
+        "LEFT JOIN team team2 ON team2.id = `match`.team2_id " +
+        "where `match`.id = ?";
+
     }
 
     let matchID = req.params.match_id;
     let matches = await db.query(sql, matchID);
-
+    if (!matches.length) {
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:577]: Don't do this as connections will drop immediately instead of waiting for event changes.

File: routes/matches/matches.js [hunk 10]
```diff
@@ -722,11 +763,14 @@ router.get("/limit/:limiter", async (req, res, next) => {
         "SELECT * FROM `match` WHERE cancelled = 0 OR cancelled IS NULL ORDER BY end_time DESC LIMIT ?";
     } else {
       sql =
-        "SELECT id, user_id, server_id, team1_id, team2_id, winner, " +
+        "SELECT `match`.id, `match`.user_id, server_id, team1_id, team2_id, winner, " +
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:766]: Same as above for styling of joins.

File: routes/matches/matches.js [hunk 11]
```diff
@@ -778,7 +822,7 @@ router.get("/limit/:limiter", async (req, res, next) => {
         "SELECT * FROM `match` WHERE cancelled = 0 OR cancelled IS NULL ORDER BY id DESC LIMIT ?,?";
     } else {
       sql =
-        "SELECT id, user_id, server_id, team1_id, team2_id, winner, " +
+        "SELECT `match`.id, user_id, server_id, team1_id, team2_id, winner, " +
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:825]: Same as above for styling of joins.

**Review by PhlexPlexico — CHANGES_REQUESTED** [review]
Hey! Thanks for taking the time to do this. However, there are some big concerns I have about bringing in data that other routes are used for, such as team names, map scores, and player stats. These should all be independent of each other, and you can do a simple API call to check and see these things. Maybe they should be changes on a front-end if you wish to see updated names.

First, the team names are stored in the match string as both a historical view of what the team was called while being played, and the only way to link a team back to a pug. Once a pug is completed, the API auto-deletes the teams and leaves that column as an only reference. If you would really want updated team names, I would recommend checking the first select and if the team does exist, then do a further select to give an updated name. 

Secondly, I see some changes to the event streams, which may be due to an older branch being used to merge this in. Please rebase on master as well, so those changes can be picked up (I highlighted a few as an example).

Thanks!

File: routes/matches/matches.js (hunk 4 above)
Comment (tech62) [reply routes/matches/matches.js:484]: Using "match" instead of `match` causing an SQL error, i will update with mtch alias as soon as possible.

Comment (PhlexPlexico) [reply routes/matches/matches.js:484]: Well, yes because match is a keyword, so you can use `FROM \`match\` as mtch` will prevent that :)

File: routes/matches/matches.js (hunk 7 above)
Comment (tech62) [reply routes/matches/matches.js:584]: I don't know how this modification happened, i've only made modifications to the SQL request.

**Comment by tech62** [comment]
Up ?

File: routes/matches/matches.js [hunk 12]
```diff
@@ -329,14 +329,19 @@ router.get("/", async (req, res, next) => {
...
       "SELECT mtch.id, mtch.user_id, mtch.server_id, mtch.team1_id, mtch.team2_id, mtch.winner, mtch.team1_score, " +
-      "mtch.team2_score, mtch.team1_series_score, mtch.team2_series_score, mtch.team1_string, mtch.team2_string, " +
+      "mtch.team2_score, mtch.team1_series_score, mtch.team2_series_score, " +
       "mtch.cancelled, mtch.forfeit, mtch.start_time, mtch.end_time, mtch.max_maps, mtch.title, mtch.skip_veto, mtch.private_match, " +
-      "mtch.enforce_teams, mtch.min_player_ready, mtch.season_id, mtch.is_pug, usr.name as owner, mp.team1_score as team1_mapscore, mp.team2_score as team2_mapscore " +
-      "FROM `match` mtch JOIN user usr ON mtch.user_id = usr.id LEFT JOIN map_stats mp ON mp.match_id = mtch.id " +
-      "WHERE cancelled = 0 " +
-      "OR cancelled IS NULL " +
+      "mtch.enforce_teams, mtch.min_player_ready, mtch.season_id, mtch.is_pug, usr.name as owner, mp.team1_score as team1_mapscore, mp.team2_score as team2_mapscore, " +
+      "gs.ip_string, gs.port, gs.gotv_port, gs.display_name, mtch.team1_string, mtch.team2_string " +
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:335]: I'm still against this query here. I don't want to be giving out game IP strings through the API without the user being authenticated. Please either break this out into an individual query (check if req.user exists, and is a part of the match/super_admin) or try not to include the IP. This gives people way too easy of a time to just DDoS a game server in drive-by attacks.

File: routes/matches/matches.js [hunk 13]
```diff
@@ -460,14 +469,28 @@ router.get("/:match_id", async (req, res, next) => {
...
-        "team1_score, team2_score, team1_series_score, team2_series_score, " +
-        "team1_string, team2_string, cancelled, forfeit, start_time, end_time, " +
-        "max_maps, title, skip_veto, private_match, enforce_teams, min_player_ready, " +
-        "season_id, is_pug, map_sides FROM `match` where id = ?";
+        "SELECT mtch.id, mtch.user_id, mtch.server_id, mtch.team1_id, mtch.team2_id, mtch.winner, " +
+        "mtch.team1_score, mtch.team2_score, mtch.team1_series_score, mtch.team2_series_score, " +
+        "mtch.cancelled, mtch.forfeit, mtch.start_time, mtch.end_time, " +
+        "mtch.max_maps, mtch.title, mtch.skip_veto, mtch.private_match, mtch.enforce_teams, mtch.min_player_ready, " +
+        "mtch.season_id, mtch.is_pug, mtch.map_sides, " +
+        "gs.ip_string, gs.port, gs.gotv_port, gs.display_name, mtch.team1_string, mtch.team2_string " +
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:490]: Also again, unprotected route, please ensure the user is logged in and has proper access to this match. If not, remove the IP.

File: routes/matches/matches.js [hunk 14]
```diff
@@ -525,14 +548,28 @@ router.get("/:match_id", async (req, res, next) => {
...
-        "team1_score, team2_score, team1_series_score, team2_series_score, " +
-        "team1_string, team2_string, cancelled, forfeit, start_time, end_time, " +
-        "max_maps, title, skip_veto, private_match, enforce_teams, min_player_ready, " +
-        "season_id, is_pug, map_sides FROM `match` where id = ?";
+        "SELECT mtch.id, mtch.user_id, mtch.server_id, mtch.team1_id, mtch.team2_id, mtch.winner, " +
+        "mtch.team1_score, mtch.team2_score, mtch.team1_series_score, mtch.team2_series_score, " +
+        "mtch.cancelled, mtch.forfeit, mtch.start_time, mtch.end_time, " +
+        "mtch.max_maps, mtch.title, mtch.skip_veto, mtch.private_match, mtch.enforce_teams, mtch.min_player_ready, " +
+        "mtch.season_id, mtch.is_pug, mtch.map_sides," +
+        "gs.ip_string, gs.port, gs.gotv_port, gs.display_name, mtch.team1_string, mtch.team2_string " +
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:569]: Same as above for SQL statement + game server IP.

File: routes/matches/matches.js [hunk 15]
```diff
@@ -719,15 +756,18 @@ router.get("/limit/:limiter", async (req, res, next) => {
     let sql;
     if (req.user !== undefined && Utils.superAdminCheck(req.user)) {
       sql =
-        "SELECT * FROM `match` WHERE cancelled = 0 OR cancelled IS NULL ORDER BY end_time DESC LIMIT ?";
+        "SELECT mtch.* FROM `match` mtch WHERE mtch.cancelled = 0 OR mtch.cancelled IS NULL ORDER BY mtch.end_time DESC LIMIT ?";
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:759]: How come this one doesn't have game server info, but the non-user statement does? Also see above for SQL statement.

File: routes/matches/matches.js [hunk 16]
```diff
@@ -775,15 +815,15 @@ router.get("/limit/:limiter", async (req, res, next) => {
     let sql;
     if (req.user !== undefined && Utils.superAdminCheck(req.user)) {
       sql =
-        "SELECT * FROM `match` WHERE cancelled = 0 OR cancelled IS NULL ORDER BY id DESC LIMIT ?,?";
+        "SELECT mtch.* FROM `match` mtch WHERE mtch.cancelled = 0 OR mtch.cancelled IS NULL ORDER BY mtch.id DESC LIMIT ?,?";
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:818]: Does this really need to be changed if you haven't made any changes to it aside from adding on the identifier when it's not needed?

File: routes/matches/matches.js [hunk 17]
```diff
@@ -775,15 +815,15 @@ router.get("/limit/:limiter", async (req, res, next) => {
...
     } else {
       sql =
-        "SELECT id, user_id, server_id, team1_id, team2_id, winner, " +
-        "team1_score, team2_score, team1_series_score, team2_series_score, " +
-        "team1_string, team2_string, cancelled, forfeit, start_time, end_time, " +
-        "max_maps, title, skip_veto, private_match, enforce_teams, min_player_ready, " +
-        "season_id, is_pug, map_sides FROM `match` WHERE cancelled = 0 " +
-        "OR cancelled IS NULL ORDER BY id DESC LIMIT ?,?";
+        "SELECT mtch.id, mtch.user_id, mtch.server_id, mtch.team1_id, mtch.team2_id, mtch.winner, " +
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:821]: Does this really need to be changed if you haven't made any changes to it aside from adding on the identifier when it's not needed?

File: routes/matches/matches.js [hunk 18]
```diff
@@ -822,7 +862,7 @@ router.get("/limit/:limiter", async (req, res, next) => {
  */
 router.get("/:match_id/config", async (req, res, next) => {
   try {
-    let sql = "SELECT * FROM `match` WHERE id = ?";
+    let sql = "SELECT mtch.* FROM `match` mtch WHERE mtch.id = ?";
```
Comment (PhlexPlexico) [inline routes/matches/matches.js:865]: Does this really need to be changed if you haven't made any changes to it aside from adding on the identifier when it's not needed?

File: routes/teams.js [hunk 19]
```diff
@@ -525,8 +526,16 @@ router.put("/", Utils.ensureAuthenticated, async (req, res) => {
     return;
   }
   let sql = "UPDATE team SET ? WHERE id=?";
+  
+  //UpdateMatchTeamName
+  let updateMatchTeam1Name = "UPDATE `match` mtch JOIN team t ON mtch.team1_id = t.id SET mtch.team1_string = t.name";
```
Comment (PhlexPlexico) [inline routes/teams.js:531]: I'm still against updating this data, as the match table shows a historical view of what a team was called. I would much prefer to check and see if the team still exists in the database, and output that name, or use the match table if the team does not exist.

File: routes/teams.js [hunk 20]
```diff
@@ -554,7 +564,15 @@ router.put("/", Utils.ensureAuthenticated, async (req, res) => {
           isCoach
         ]);
       }
+
```
Comment (PhlexPlexico) [inline routes/teams.js:567]: Remove whitespace.

**Review by PhlexPlexico — CHANGES_REQUESTED** [review]
Not sure if you deleted your comment, but a PUG is a Pick Up Game, or a 10 man, etc. When you get 10 people and assign random teams. Not much use for a tournament, but good for when people want to track stats externally when they play.

As for these changes, there are still a handful of concerns I have, mainly for leaking game server IPs when a user is not authenticated on the API. This would cause a myriad of issues, and breaks the security logic in the application (why go to `/servers` and forced to be logged in when I can enumerate all the servers that have been actively used with `/matches`).

As I had said previously as well, I'm still against changing the Team names in the match table, as they provide a temporal list of what teams were called. I think the best bet would be to adjust the select statement and show if the same team ID still exists in the teams table, and if it doesn't, then show the team1_string and team2_string respectively.

Lastly, this is by no means my full time job, this is a hobby. There is no need to ping or update the PR to get my attention. I'll get to it once I have time.

Thanks!

**Comment by PhlexPlexico** [comment]
Since there hasn't been any updates to this in 7 months, I'm going to mark this as closed. Feel free to reopen and fix the changes that were listed/reasoning behind some of them (i.e. leaking IPs of servers publically).

Thanks!
